services/graph/pkg/service/v0/api_driveitem_permissions.go

```diff
@@ -0,0 +1,669 @@
+		return errorcode.New(errorcode.ItemNotFound, "permission not found")
+	case sharedResourceID == nil:
+		return errorcode.New(errorcode.ItemNotFound, "failed to resolve resource id for shared resource")
+	}
+
+	// The resourceID of the shared resource need to match the item ID from the Request Path
+	// otherwise this is an invalid Request.
+	if !utils.ResourceIDEqual(sharedResourceID, &itemID) {
+		s.logger.Debug().Msg("resourceID of shared does not match itemID")
+		return errorcode.New(errorcode.InvalidRequest, "permissionID and itemID do not match")
+	}
+
+	switch permissionType {
+	case User:
+		return s.removeUserShare(ctx, permissionID)
+	case Public:
+		return s.removePublicShare(ctx, permissionID)
+	case Space:
+		return s.removeSpacePermission(ctx, permissionID, sharedResourceID)
+	}
+
+	// This should never be reached
+	return errorcode.New(errorcode.GeneralException, "failed to delete permission")
+}
+
+func (s DriveItemPermissionsService) DeleteSpaceRootPermission(ctx context.Context, driveID storageprovider.ResourceId, permissionID string) error {
+	gatewayClient, err := s.gatewaySelector.Next()
+	if err != nil {
+		return err
+	}
+
+	space, err := utils.GetSpace(ctx, storagespace.FormatResourceID(driveID), gatewayClient)
+	if err != nil {
+		return err
+	}
+
+	if space.SpaceType != _spaceTypeProject {
+		return errorcode.New(errorcode.InvalidRequest, "unsupported space type")
+	}
+
+	rootResourceID := space.GetRoot()
+	return s.DeletePermission(ctx, *rootResourceID, permissionID)
+}
+
+func (s DriveItemPermissionsService) UpdatePermission(ctx context.Context, itemID storageprovider.ResourceId, permissionID string, newPermission libregraph.Permission) (libregraph.Permission, error) {
+	oldPermission, sharedResourceID, err := s.getPermissionByID(ctx, permissionID, &itemID)
+	if err != nil {
+		return libregraph.Permission{}, err
+	}
+
+	// The resourceID of the shared resource need to match the item ID from the Request Path
+	// otherwise this is an invalid Request.
+	if !utils.ResourceIDEqual(sharedResourceID, &itemID) {
+		s.logger.Debug().Msg("resourceID of shared does not match itemID")
+		return libregraph.Permission{}, errorcode.New(errorcode.InvalidRequest, "permissionID and itemID do not match")
+	}
+
+	// This is a public link
+	if _, ok := oldPermission.GetLinkOk(); ok {
+		updatedPermission, err := s.updatePublicLinkPermission(ctx, permissionID, &itemID, &newPermission)
+		if err != nil {
+			return libregraph.Permission{}, err
+		}
+		return *updatedPermission, nil
+	}
+
+	// This is a user share
+	updatedPermission, err := s.updateUserShare(ctx, permissionID, sharedResourceID, &newPermission)
+	if err != nil {
+		return libregraph.Permission{}, err
+	}
+	return *updatedPermission, nil
+}
+
+func (s DriveItemPermissionsService) UpdateSpaceRootPermission(ctx context.Context, driveID storageprovider.ResourceId, permissionID string, newPermission libregraph.Permission) (libregraph.Permission, error) {
```

Comment:
exported func `UpdateSpaceRootPermission` should have a comment or be unexported